app/scripts/ui.js

```diff
@@ -60,23 +60,50 @@ async function start() {
    * Code below ensures that UI is rendered only after background is ready.
    */
   if (isManifestV3()) {
-    extensionPort.onMessage.addListener((message) => {
+    /*
+     * In case of MV3 the issue of blank screen was very frequent, it is caused by UI initialising before background is ready to send state.
+     * Code below ensures that UI is rendered only after CONNECTION_READY message is received thus background is ready.
+     * In case UI is already renderd only stream will be updated.
+     */
+    const messageListener = (message) => {
       if (message?.name === 'CONNECTION_READY') {
-        initializeUiWithTab(activeTab);
+        if (isUIInitialised) {
+          updateUiStreams();
+        } else {
+          initializeUiWithTab(activeTab);
+        }
       }
-    });
+    };
+
+    // disconnectListener takes care to remove listeners from closed streams
+    // it also creates new streams and attach event listeners to them
+    const disconnectListener = () => {
+      extensionPort.onMessage.removeListener(messageListener);
+      extensionPort.onDisconnect.removeListener(disconnectListener);
+      // message below will try to activate service worker
+      // in MV3 is likely that reason of stream closing is service worker going in-active
+      browser.runtime.sendMessage({ name: 'UI_OPEN' });
```

Comment:
is this message specifically listened for anywhere?

Reply:
This is captured by service worker and is used to activate it in case it becomes in-active.

Reply:
[Here](https://github.com/MetaMask/metamask-extension/pull/14781/files#diff-cd37b653f9862ccada12e440c6ef67e8b2b2c5b906a320870ed71232d244e660R63)? Should we specifically listen for this message?

Reply:
Hey @adonesky1 : sorry for delay, it seems link above does not works. Can you plz update.